pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -0,0 +1,272 @@
+package planner
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	rkev1 "github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1"
+	"github.com/rancher/rancher/pkg/apis/rke.cattle.io/v1/plan"
+	"github.com/rancher/rancher/pkg/controllers/provisioningv2/rke2"
+	ranchercontrollers "github.com/rancher/rancher/pkg/generated/controllers/provisioning.cattle.io/v1"
+	rkecontrollers "github.com/rancher/rancher/pkg/generated/controllers/rke.cattle.io/v1"
+	"github.com/rancher/rancher/pkg/provisioningv2/kubeconfig"
+	"github.com/rancher/rancher/pkg/wrangler"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type encryptionKeyRotation struct {
+	kubeconfigManager   *kubeconfig.Manager
+	rancherClusterCache ranchercontrollers.ClusterCache
+	rkeControlPlanes    rkecontrollers.RKEControlPlaneController // convert to controller in order to reenqueue
+	store               *PlanStore
+}
+
+func newEncryptionKeyRotation(clients *wrangler.Context, store *PlanStore, kubeconfigManager *kubeconfig.Manager, rancherClusterCache ranchercontrollers.ClusterCache) *encryptionKeyRotation {
+	return &encryptionKeyRotation{
+		kubeconfigManager:   kubeconfigManager,
+		rancherClusterCache: rancherClusterCache,
+		rkeControlPlanes:    clients.RKE.RKEControlPlane(),
+		store:               store,
+	}
+}
+
+func (r *encryptionKeyRotation) RotateEncryptionKeys(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan) error {
+	if !shouldRotateKeys(cp) {
+		return nil
+	}
+
+	if shouldResetEncryptionStatus(cp) {
+		cp.Status.RotateEncryptionKeysPhase = rkev1.RotateEncryptionKeysPhaseStart
+		_, err := r.rkeControlPlanes.UpdateStatus(cp)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	currentInitNodes := collect(clusterPlan, isInitNode)
```

Comment:
You shouldn't need to do this, there is a helper function to find the init node in `initnode.go` of the planner